src/Interpreters/GinFilter.cpp

```diff
@@ -0,0 +1,224 @@
+#include <string>
+#include <algorithm>
+#include <city.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnNullable.h>
+#include <Columns/ColumnLowCardinality.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <Storages/MergeTree/MergeTreeIndexGin.h>
+#include <Storages/MergeTree/MergeTreeIndexFullText.h>
+#include <Disks/DiskLocal.h>
+#include <Storages/MergeTree/GinIndexStore.h>
+#include <Interpreters/GinFilter.h>
+
+using namespace std;
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+GinFilterParameters::GinFilterParameters(size_t ngrams_)
+    : ngrams(ngrams_)
+{
+    if (ngrams > 8)
+        throw Exception("The size of gin filter cannot be greater than 8", ErrorCodes::BAD_ARGUMENTS);
+}
+
+GinFilter::GinFilter(const GinFilterParameters & params)
+    : ngrams(params.ngrams)
+{
+}
+
+void GinFilter::add(const char* data, size_t len, UInt32 rowID, GinIndexStorePtr& store)
+{
+    if (len > MAX_TERM_LENGTH)
+        return;
+
+    string token(data, len);
```

Comment:
Realized you need to preserve memory, then create string only on line 51

Reply:
Since std::string_view needs underlying data having longer lifetime, we have to use string here. But it is a good point for a larger scale of refactoring in the future.